02_bubble/jecofang/sort.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	n := len(s)
+	arr := make([]int, n)
+	copy(arr, s)
+
+	if n < 2 {
+		return arr
+	}
+
+	swap := true
+	for swap {
+		swap = false
+		for i := 0; i < n-1; i++ {
+			if arr[i+1] < arr[i] {
+				arr[i], arr[i+1] = arr[i+1], arr[i]
+				swap = true
+			}
+		}
+	}
+
+	return arr
+}
+
+func insertion(s []int) []int {
+	n := len(s)
+	arr := make([]int, n)
+	copy(arr, s)
+
+	if n < 2 {
+		return arr
+	}
+
+	for i := 1; i < n; i++ {
+		for j := i; j > 0; j-- {
+			if arr[j] < arr[j-1] {
+				arr[j-1], arr[j] = arr[j], arr[j-1]
+			}
+		}
+	}
+
+	return arr
+}
+
+func quick(s []int) []int {
+	n := len(s)
+
+	if n < 2 {
+		arr := make([]int, n)
+		copy(arr, s)
+		return arr
+	}
+
+	left := make([]int, 0, n)
```

Comment:
use many memory here. Try swapping on the slice.

Reply:
Good point, @linwumeng - but then @jecofang01 you should probably add a comment "sorts []int in place" - or you need a wrapping function that first creates a copy or the original array.